File: lib/ui/onboarding/face_scan_ring.dart
import 'dart:math' as math;

import 'package:camera/camera.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

import '../../theme/app_theme.dart';

/// 人脸录入结果态（驱动扫描环遮罩与配色）。
enum FaceScanState { idle, collecting, success, failed, duplicate }

/// Face-ID 风格的人脸扫描环组件（圆形摄像头取景框 + 放射状刻度环）。
///
/// 从 `face_registration_screen.dart` 抽取的共享组件，向导与设置页录入共用。
/// 通过 [cameraController] 直接预览常驻主相机流（不另开 CameraController），
/// 与主识别管线共享同一帧，避免坐标空间错位。
///
/// 组件本身不负责采集逻辑（由调用方驱动 `captureFaceSample`），只负责展示：
/// 扫描动画 [sweep]、采集进度 [progress]、结果态 [state]。
class FaceScanRing extends StatelessWidget {
  const FaceScanRing({
    super.key,
    required this.cameraController,
    required this.state,
    required this.progress,
    required this.sweep,
    this.diameter,
  });

  /// 常驻主相机控制器（AppController.cameraController）。
  final CameraController? cameraController;

  /// 当前采集状态。
  final FaceScanState state;

  /// 采集进度 [0,1]（已采集 / 需采集）。
  final double progress;

  /// 扫描动画当前值 [0,1]，null 表示静止。
  final double? sweep;

  /// 直径；为空时由父级约束决定。
  final double? diameter;

  static const Color _activeColor = AppTheme.accent;
  static const Color _baseColor = Color(0xFF38383A);
  static const Color _card = Color(0xFF1C1C1E);
  static const Color _green = AppTheme.accentGreen;
  static const Color _red = AppTheme.accentRed;

  Color get _ringColor {
    switch (state) {
      case FaceScanState.success:
        return _green;
      case FaceScanState.failed:
      case FaceScanState.duplicate:
        return _red;
      default:
        return _activeColor;
    }
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (context, constraints) {
        final maxWidth = constraints.maxWidth;
        final maxHeight = constraints.maxHeight;
        final size = (diameter ??
                math.min(maxWidth, maxHeight))
            .clamp(96.0, 360.0)
            .toDouble();
        return Center(child: _buildCircle(size));
      },
    );
  }

  Widget _buildCircle(double size) {
    const ringPadding = 22.0;
    final previewSize = size - ringPadding * 2;

    return SizedBox(
      width: size,
      height: size,
      child: Stack(
        alignment: Alignment.center,
        children: [
          CustomPaint(
            size: Size(size, size),
            painter: _FaceIdRingPainter(
              activeColor: _ringColor,
              baseColor: _baseColor,
              sweep: sweep,
              progress: progress,
            ),
          ),
          ClipOval(
            child: SizedBox(
              width: previewSize,
              height: previewSize,
              child: _buildPreview(previewSize),
            ),
          ),
          if (state == FaceScanState.success)
            Container(
              width: previewSize,
              height: previewSize,
              decoration: BoxDecoration(
                shape: BoxShape.circle,
                color: _green.withValues(alpha: 0.18),
              ),
              child: const Icon(CupertinoIcons.checkmark_alt,
                  color: _green, size: 72),
            ),
          if (state == FaceScanState.failed ||
              state == FaceScanState.duplicate)
            Container(
              width: previewSize,
              height: previewSize,
              decoration: BoxDecoration(
                shape: BoxShape.circle,
                color: _red.withValues(alpha: 0.18),
              ),
              child: const Icon(CupertinoIcons.exclamationmark,
                  color: _red, size: 64),
            ),
        ],
      ),
    );
  }

  Widget _buildPreview(double size) {
    final controller = cameraController;
    if (controller == null || !controller.value.isInitialized) {
      return Container(
        color: _card,
        child: const Center(
          child:
              CupertinoActivityIndicator(color: Colors.white, radius: 14),
        ),
      );
    }
    // 强制横向比例（宽>高），由 FittedBox.cover 按长边裁剪铺满圆形，避免拉伸。
    final preview = controller.value.previewSize ?? Size(size, size);
    final maxSide = math.max(preview.width, preview.height);
    final minSide = math.min(preview.width, preview.height);
    return ClipRect(
      child: OverflowBox(
        alignment: Alignment.center,
        child: FittedBox(
          fit: BoxFit.cover,
          child: SizedBox(
            width: maxSide,
            height: minSide,
            child: CameraPreview(controller),
          ),
        ),
      ),
    );
  }
}

/// Face ID 风格的放射状刻度环。
class _FaceIdRingPainter extends CustomPainter {
  _FaceIdRingPainter({
    required this.activeColor,
    required this.baseColor,
    this.sweep,
    this.progress = 0,
  });

  static const int tickCount = 72;
  final Color activeColor;
  final Color baseColor;
  final double? sweep;

  /// 采集进度 [0,1]，决定已完成刻度的染色比例。
  final double progress;

  @override
  void paint(Canvas canvas, Size size) {
    final center = Offset(size.width / 2, size.height / 2);
    final outerRadius = size.width / 2 - 2;
    const baseLen = 7.0;
    const activeLen = 14.0;

    for (var i = 0; i < tickCount; i++) {
      final t = i / tickCount;
      final angle = -math.pi / 2 + 2 * math.pi * t;
      final len = sweep != null ? activeLen : baseLen;
      final inner = outerRadius - len;
      final cosA = math.cos(angle);
      final sinA = math.sin(angle);
      final p1 = Offset(center.dx + cosA * inner, center.dy + sinA * inner);
      final p2 = Offset(
          center.dx + cosA * outerRadius, center.dy + sinA * outerRadius);

      var color = baseColor;
      var width = 2.0;

      // 已采集部分整体染色，给予进度感。
      if (t < progress) {
        color = Color.lerp(baseColor, activeColor, 0.55)!;
        width = 2.5;
      }

      // 录入中：流动高光环绕。
      if (sweep != null) {
        final dist = (sweep! - t).abs();
        final glow = (1 - (dist * 6)).clamp(0.0, 1.0);
        if (glow > 0) {
          color = Color.lerp(color, activeColor, glow)!;
          width = 3.0 + glow * 1.5;
        }
      }
      canvas.drawLine(p1, p2,
          Paint()..color = color..strokeWidth = width..strokeCap = StrokeCap.round);
    }
  }

  @override
  bool shouldRepaint(covariant _FaceIdRingPainter old) =>
      old.activeColor != activeColor ||
      old.baseColor != baseColor ||
      old.sweep != sweep ||
      old.progress != progress;
}

File: xbot-native/app/src/main/java/com/xbot/android/ui/FaceScanRing.kt
package com.xbot.android.ui

import androidx.camera.core.ImageProxy
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.lerp
import androidx.compose.ui.unit.dp
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.PriorityHigh
import java.util.concurrent.Executor
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.sin

/** 人脸录入结果态（驱动扫描环遮罩与配色）。对应 Flutter FaceScanState。 */
enum class FaceScanState { IDLE, COLLECTING, SUCCESS, FAILED, DUPLICATE }

private val RingAccent = Color(0xFF0A84FF) // iOS systemBlue
private val RingBase = Color(0xFF38383A)
private val RingGreen = Color(0xFF30D158)
private val RingRed = Color(0xFFFF453A)
private val PreviewCard = Color(0xFF1C1C1E)

/**
 * Face-ID 风格的人脸扫描环（圆形摄像头取景 + 放射状刻度环）。对应 Flutter FaceScanRing。
 *
 * - 圆形相机预览（TextureView 兼容模式裁剪为圆）。
 * - 72 根放射刻度：采集进度染色 + 录入时流动高光环绕。
 * - 结果态：成功绿色对勾遮罩 / 失败·重复红色叹号遮罩。
 *
 * @param progress 采集进度 0..1
 * @param scanning 是否正在录入（驱动高光环绕动画 + 加长刻度）
 * @param cameraActive 相机是否就绪（false 显示加载圈）
 */
@Composable
fun FaceScanRing(
    state: FaceScanState,
    progress: Float,
    scanning: Boolean,
    cameraActive: Boolean,
    analysisExecutor: Executor,
    analyzer: (ImageProxy) -> Unit,
    modifier: Modifier = Modifier,
) {
    val ringColor = when (state) {
        FaceScanState.SUCCESS -> RingGreen
        FaceScanState.FAILED, FaceScanState.DUPLICATE -> RingRed
        else -> RingAccent
    }

    val transition = rememberInfiniteTransition(label = "faceScanSweep")
    val anim by transition.animateFloat(
        initialValue = 0f,
        targetValue = 1f,
        animationSpec = infiniteRepeatable(
            animation = tween(2000, easing = LinearEasing),
            repeatMode = RepeatMode.Restart,
        ),
        label = "sweep",
    )
    val sweep: Float? = if (scanning) anim else null

    BoxWithConstraints(modifier = modifier, contentAlignment = Alignment.Center) {
        val side = min(maxWidth.value, maxHeight.value).coerceIn(180f, 360f).dp
        val ringPadding = 22.dp
        val previewSide = side - ringPadding * 2

        Box(modifier = Modifier.size(side), contentAlignment = Alignment.Center) {
            // 放射刻度环。
            Canvas(modifier = Modifier.size(side)) {
                val center = Offset(size.width / 2f, size.height / 2f)
                val outer = size.minDimension / 2f - 2.dp.toPx()
                val baseLen = 7.dp.toPx()
                val activeLen = 14.dp.toPx()
                val tickCount = 72
                for (i in 0 until tickCount) {
                    val t = i / tickCount.toFloat()
                    val angle = -PI / 2 + 2 * PI * t
                    val len = if (sweep != null) activeLen else baseLen
                    val inner = outer - len
                    val ca = cos(angle).toFloat()
                    val sa = sin(angle).toFloat()
                    val p1 = Offset(center.x + ca * inner, center.y + sa * inner)
                    val p2 = Offset(center.x + ca * outer, center.y + sa * outer)
                    var color = RingBase
                    var width = 2.dp.toPx()
                    if (t < progress) {
                        color = lerp(RingBase, ringColor, 0.55f)
                        width = 2.5.dp.toPx()
                    }
                    if (sweep != null) {
                        val dist = kotlin.math.abs(sweep - t)
                        val glow = (1f - dist * 6f).coerceIn(0f, 1f)
                        if (glow > 0f) {
                            color = lerp(color, ringColor, glow)
                            width = (3f + glow * 1.5f).dp.toPx()
                        }
                    }
                    drawLine(color = color, start = p1, end = p2, strokeWidth = width, cap = StrokeCap.Round)
                }
            }

            // 圆形相机预览。
            Box(
                modifier = Modifier
                    .size(previewSide)
                    .clip(CircleShape)
                    .background(PreviewCard),
                contentAlignment = Alignment.Center,
            ) {
                if (cameraActive) {
                    CameraPreview(
                        modifier = Modifier.fillMaxSize().clip(CircleShape),
                        useFront = true,
                        active = true,
                        compatibleMode = true,
                        analysisExecutor = analysisExecutor,
                        analyzer = analyzer,
                    )
                } else {
                    CircularProgressIndicator(color = Color.White)
                }

                // 结果遮罩。
                when (state) {
                    FaceScanState.SUCCESS -> ResultOverlay(RingGreen, Icons.Filled.Check)
                    FaceScanState.FAILED, FaceScanState.DUPLICATE ->
                        ResultOverlay(RingRed, Icons.Filled.PriorityHigh)
                    else -> {}
                }
            }
        }
    }
}

@Composable
private fun ResultOverlay(color: Color, icon: androidx.compose.ui.graphics.vector.ImageVector) {
    Box(
        modifier = Modifier
            .fillMaxSize()
            .clip(CircleShape)
            .background(color.copy(alpha = 0.18f)),
        contentAlignment = Alignment.Center,
    ) {
        Icon(
            imageVector = icon,
            contentDescription = null,
            tint = color,
            modifier = Modifier.size(72.dp),
        )
    }
}
